Add build, flash, serial and troubleshooting tutorial for lab02 (#50)
UPB-FILS-MA/upb-fils-ma.github.io: docs/lab/02/02.md

:::

## Build and flash

### Build

> To build a program for the Raspberry Pi Pico, run the following command (in the project root folder):

Reviewer: ```suggestion
To build a program for the Raspberry Pi Pico, run the following command in the crate's root folder (where the `Cargo.toml` file and `src` folder are):
```

> `cargo build --bin <project_name>`

Reviewer: ```suggestion
```shell
cargo build
```
```

We have different crates now for the tasks.

> where `<project_name>` is to be replaced with the name of your project.

Reviewer: ```suggestion
```

> This will create the binary file inside `/target/thumbv6m-none-eabi/debug/`.

Reviewer: ```suggestion
This will create the binary file inside `target/thumbv6m-none-eabi/debug/`.

:::note

If the crate is part of a workspace (as it is the lab template), the `target` folder is located in the workspace's root instead of the crate's root.

:::
```

### Flash

To load your program to the Raspberry Pi Pico, connect the board through USB to your computer while holding down the `BOOTSEL` button. It should appear as an external drive on your system. Afterwards, you can run the following command (also from the root folder):

> `elf2uf2-rs -d .\target\thumbv6m-none-eabi\debug\<project_name>`

Reviewer: ```suggestion
```shell
elf2uf2-rs -d target/thumbv6m-none-eabi/debug/<crate_name>
```
```

The / delimiter works for both windows and linux systems.

Author: Changed all paths.

> 

Reviewer: ```note

In the lab template, the `target` folder is located two levels up relative to the lab's crates. This means the path is `../../target/thumbv6m-none-eabi/debug/<crate_name>`

```

Author: Added a note pointing this out to students.

This will upload your program to the Pico.

:::note

If the above command doesn't work, try these steps instead:

> - run this command: `elf2uf2-rs .\target\thumbv6m-none-eabi\debug\<project_name>`

Reviewer: ```suggestion
- run this command: 
```shell
elf2uf2-rs target/thumbv6m-none-eabi/debug/<crate_name>
```
```

> - drag and drop the `.uf2` file located in the `.\target\thumbv6m-none-eabi\debug\` folder to the RP external drive

Reviewer: ```suggestion
- drag and drop the `.uf2` file located in the `target/thumbv6m-none-eabi/debug/` folder to the RP external drive
```

:::

### Serial console

To see messages from the Pico over the serial port, add the `-s` argument to the flash command:

> `elf2uf2-rs -s -d .\target\thumbv6m-none-eabi\debug\<project_name>`

Reviewer: ```suggestion
```shell
elf2uf2-rs -s -d target/thumbv6m-none-eabi/debug/<crate_name>
```
```

This will allow us to see messages sent over serial from the board.

### Troubleshooting

In case your code doesn't seem to get uploaded correctly or the board goes into `BOOTSEL` mode as soon as you plug it into your computer, it might be time to debug. To install required dependencies for debugging, run these commands:

> `rustup component add llvm-tools`
> 
> `cargo install cargo-binutils`

Reviewer: ```suggestion
```shell
rustup component add llvm-tools
cargo install cargo-binutils
```
```

Author: Modified and added correct shell code symbol.

Now, you can run this command:

> `rust-objdump --section-headers .\target\thumbv6m-none-eabi\debug\<project_name>`

Reviewer: ```suggestion
```shell
rust-objdump --section-headers target/thumbv6m-none-eabi/debug/<crate_name>
```
```

```shell
cargo build

This will create the binary file inside `target/thumbv6m-none-eabi/debug/`.

:::note

If the crate is part of a workspace (as it is the lab template), the `target` folder is located in the workspace's root instead of the crate's root.

:::

### Flash

To load your program to the Raspberry Pi Pico, connect the board through USB to your computer while holding down the `BOOTSEL` button. It should appear as an external drive on your system. Afterwards, you can run the following command (also from the root folder):

`elf2uf2-rs -d .\target\thumbv6m-none-eabi\debug\<project_name>`

This will upload your program to the Pico.

:::note

If the above command doesn't work, try these steps instead:

- run this command: `elf2uf2-rs .\target\thumbv6m-none-eabi\debug\<project_name>`

- drag and drop the `.uf2` file located in the `.\target\thumbv6m-none-eabi\debug\` folder to the RP external drive

:::

### Serial console

To see messages from the Pico over the serial port, add the `-s` argument to the flash command:

`elf2uf2-rs -s -d .\target\thumbv6m-none-eabi\debug\<project_name>`

This will allow us to see messages sent over serial from the board.

### Troubleshooting

> 

Reviewer: ```suggestion

#### Link error - reset vector is missing

#### Flash succeeds, the application does not start

```

Author: Added the two separate sections, and information on the reset vector error.
